WIP: Add ProgressPreference to SpecialVariables var and type arrays

cmkb3 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

<!-- Summarize your PR between here and the checklist. -->
This fixes an array out of bounds/type error that occurs when using the -ProgressAction common parameter in advanced functions. The reference variable and variable type arrays are missing the ProgressPreference entry which causes various engine errors including index out of bounds and invalid typecast from index collision.

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->
Fixes: #21074 'The -ProgressAction parameter doesn't work (PowerShell 7.4.1 only)'
Fixes: #20657 'The new -ProgressAction common parameter breaks calls to advanced scripts and functions. '


## PR Checklist

- [✅] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [✅] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [✅] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
  - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [✅] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
  - [✅] Not Applicable
  - **OR**
  - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
  - [ ] N/A or can only be tested interactively
  - **OR**
  - [✅] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
  - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
  - **OR**
  - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
    - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
      - [ ] Issue filed: <!-- Number/link of that issue here -->

cmkb3 committed:
Update SpecialVariables.cs

jborean93 committed:
Added tests

cmkb3 commented:
> @cmkb3 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

pwshBot committed:
Bump .NET 9 and dependencies (#24573)

cmkb3 committed:
Merge branch 'PowerShell:master' into ProgressPreference_Fix

iSazonov commented on test/powershell/Language/Scripting/CommonParameters.Tests.ps1:
```diff
@@ -147,6 +147,40 @@ Describe "Common parameters support for script cmdlets" -Tags "CI" {
     }
 
+    Context "ProgressAction" {
+        It "Silences with no parameters/variables" {
```
Please reword tests names. "Silence" is not explain that test checks.

iSazonov reviewed (commented):
Thanks for finding and patching the issue.

I suggest to discuss replacing `Progress = 16,` with `Progress = 8,` in `internal enum PreferenceVariable` for performance reason (with the PR the tuple size jumps from 16 to 32).

cmkb3 commented:
> Thanks for finding and patching the issue.
> 
> I suggest to discuss replacing `Progress = 16,` with `Progress = 8,` in `internal enum PreferenceVariable` for performance reason (with the PR the tuple size jumps from 16 to 32).

If I understand correctly, the Enum value must be unique within the combination of both `AutomaticVariable` and `PreferenceVariable` internal enum sets. Value 8 already exists within `AutomaticVariable` as `PSCommandPath`.

    internal enum AutomaticVariable
     {
        Underbar = 0,
        Args = 1,
        This = 2,
        Input = 3,
        PSCmdlet = 4,
        PSBoundParameters = 5,
        MyInvocation = 6,
        PSScriptRoot = 7,
        PSCommandPath = 8,
        NumberOfAutomaticVariables // 1 + the last, used to initialize global scope.
    }

    internal enum PreferenceVariable
    {
        Debug = 9,
        Verbose = 10,
        Error = 11,
        WhatIf = 12,
        Warning = 13,
        Information = 14,
        Confirm = 15,
        Progress = 16,
    }

iSazonov commented:
> If I understand correctly,

Yes, I see. Could you please add new Assert like in line 718 for `PreferenceVariableTypes`?

If we can not use value less 16 we will allocate double size tuples so we have to think about whether we need to fix the performance issue and how.

jborean93 commented:
> If we can not use value less 16 we will allocate double size tuples so we have to think about whether we need to fix the performance issue and how.

While we should think about this the damage is already done by accepting and implementing the new parameter in 7.4 . We should fix this issue first then investigate the performance side after not hold off a bugfix for a problem today.

cmkb3 commented:
> Yes, I see. Could you please add new Assert like in line 718 for `PreferenceVariableTypes`?

What is this in reference to? Neither of the changed files are above 500 lines; searching for 'Assert' in either returned no results for me.

iSazonov commented:
> > Yes, I see. Could you please add new Assert like in line 718 for `PreferenceVariableTypes`?
> 
> What is this in reference to? Neither of the changed files are above 500 lines; searching for 'Assert' in either returned no results for me.

Ah, sorry! https://github.com/PowerShell/PowerShell/blob/c2f1ff00ae43bc81b4e26168792a1e1118c0d4ed/src/System.Management.Automation/engine/parser/VariableAnalysis.cs#L718-L721


> > If we can not use value less 16 we will allocate double size tuples so we have to think about whether we need to fix the performance issue and how.
> 
> While we should think about this the damage is already done by accepting and implementing the new parameter in 7.4 . We should fix this issue first then investigate the performance side after not hold off a bugfix for a problem today.

Currently we haven't the performance issue because the feature doesn't work. We should either revert https://github.com/PowerShell/PowerShell/pull/18887 or fix without any (perf) regression.

jborean93 commented:
> Currently we haven't the performance issue because the feature doesn't work. We should either revert https://github.com/PowerShell/PowerShell/pull/18887 or fix without any (perf) regression.

It works just fine for binary cmdlets, it’s advanced functions where it fails and this fixes. We only loose out on the size 16 tuple if the function didn’t have any paramaters or variables (or was un-optimisable). I would say the vast majority of advanced functions will have at least one parameter or one variable making it go beyond 16 in any case.

cmkb3 commented:
> > > Yes, I see. Could you please add new Assert like in line 718 for `PreferenceVariableTypes`?
> > 
> > 
> > What is this in reference to? Neither of the changed files are above 500 lines; searching for 'Assert' in either returned no results for me.
> 
> Ah, sorry!
> 
> https://github.com/PowerShell/PowerShell/blob/c2f1ff00ae43bc81b4e26168792a1e1118c0d4ed/src/System.Management.Automation/engine/parser/VariableAnalysis.cs#L718-L721
> 
> > > If we can not use value less 16 we will allocate double size tuples so we have to think about whether we need to fix the performance issue and how.
> > 
> > 
> > While we should think about this the damage is already done by accepting and implementing the new parameter in 7.4 . We should fix this issue first then investigate the performance side after not hold off a bugfix for a problem today.
> 
> Currently we haven't the performance issue because the feature doesn't work. We should either revert #18887 or fix without any (perf) regression.

Are you looking for something like this? Asserts that the enums match the associated array names/types lengths.

```cs
Diagnostics.Assert(AutomaticVariable.NumberOfAutomaticVariables == SpecialVariables.AutomaticVariables.Length
    && AutomaticVariable.NumberOfAutomaticVariables == SpecialVariables.AutomaticVariableTypes.Length,
    "automatic variable enum length does not match both automatic variable and automatic variable types length.");

Diagnostics.Assert(SpecialVariables.PreferenceVariable.Length == SpecialVariables.PreferenceVariables.Length
    && SpecialVariables.PreferenceVariable.Length == SpecialVariables.PreferenceVariableTypes.Length,
    "preference variable enum length does not match both preference variable and preference variable types length.").
```

iSazonov commented:
> Are you looking for something like this (@iSazonov)? Asserts that the enums match the associated array names/types lengths.

Yes. We should add check for types you change in the PR.

Have you checked the existing assert in DEBUG mode? I guess it will throw now.

adityapatwardhan committed:
Update machine pool for copy blob and upload buildinfo stage (#24587)

jborean93 commented:
> Have you checked the existing assert in DEBUG mode? I guess it will throw now.

The debug statements

```
Diagnostics.Assert(AutomaticVariable.NumberOfAutomaticVariables == SpecialVariables.AutomaticVariables.Length
    && AutomaticVariable.NumberOfAutomaticVariables == SpecialVariables.AutomaticVariableTypes.Length,
    "automatic variable enum length does not match both automatic variable and automatic variable types length.");

Diagnostics.Assert(SpecialVariables.PreferenceVariable.Length == SpecialVariables.PreferenceVariables.Length
    && SpecialVariables.PreferenceVariable.Length == SpecialVariables.PreferenceVariableTypes.Length,
    "preference variable enum length does not match both preference variable and preference variable types length.").
```

Do not throw as they are just checking that the slots used for the automatic vars and action preferences match up together. In this case the enum wasn't populated with the progress preference var so the slots still match up with what was in the enum.

If we wanted to add a debug assertion we could potentially do add something to the below which somehow checks the enums defined [here](https://github.com/PowerShell/PowerShell/blob/c2f1ff00ae43bc81b4e26168792a1e1118c0d4ed/src/System.Management.Automation/engine/SpecialVariables.cs#L413-L423). I don't know off a good check for this off the top of my head though.

https://github.com/PowerShell/PowerShell/blob/c2f1ff00ae43bc81b4e26168792a1e1118c0d4ed/src/System.Management.Automation/engine/parser/VariableAnalysis.cs#L154-L159

cmkb3 commented:
> The debug statements
> 
> [...]
>
> Do not throw as they are just checking that the slots used for the automatic vars and action preferences match up together. In this case the enum wasn't populated with the progress preference var so the slots still match up with what was in the enum.

If the second debug statement I proposed existed in the current codebase (pre-merge), it should have thrown, if I understand correctly.

```cs
// Enum, length of 8
internal enum PreferenceVariable
{
    Debug = 9,
    Verbose = 10,
    Error = 11,
    WhatIf = 12,
    Warning = 13,
    Information = 14,
    Confirm = 15,
    Progress = 16,
};

// Array of preference values, length of 7
internal static readonly string[] PreferenceVariables =
{
    SpecialVariables.DebugPreference,
    SpecialVariables.VerbosePreference,
    SpecialVariables.ErrorActionPreference,
    SpecialVariables.WhatIfPreference,
    SpecialVariables.WarningPreference,
    SpecialVariables.InformationPreference,
    SpecialVariables.ConfirmPreference,
};

// Array of preference variable types, length of 7
internal static readonly string[] PreferenceVariables =
{
    SpecialVariables.DebugPreference,
    SpecialVariables.VerbosePreference,
    SpecialVariables.ErrorActionPreference,
    SpecialVariables.WhatIfPreference,
    SpecialVariables.WarningPreference,
    SpecialVariables.InformationPreference,
    SpecialVariables.ConfirmPreference,
};

// Debug assert, effectively becomes "Assert 8 == 7 && 8 == 7"
Diagnostics.Assert(SpecialVariables.PreferenceVariable.Length == SpecialVariables.PreferenceVariables.Length
    && SpecialVariables.PreferenceVariable.Length == SpecialVariables.PreferenceVariableTypes.Length,
    "preference variable enum length does not match both preference variable and preference variable types length.").
    ```

jborean93 commented:
Ah my apologies I misread things.

cmkb3 committed:
Update VariableAnalysis.cs

Add assertion to prevent automatic or preference variable enums from being misaligned with their associated string and type arrays

cmkb3 committed:
Update VariableAnalysis.cs

Add assertion to prevent automatic or preference variable enums from being misaligned with their associated string and type arrays

cmkb3 committed:
Merge branch 'PowerShell:master' into ProgressPreference_Fix

cmkb3 committed:
Update VariableAnalysis.cs

cmkb3 committed:
Update VariableAnalysis.cs

cmkb3 committed:
Fix introduced typos in VariableAnalysis Asserts

cmkb3 committed:
Fix introduced typos in VariableAnalysis Asserts

cmkb3 committed:
Update CommonParameter ProgressAction Pester test names, add manual Write-Progress inside to verify no output

cmkb3 committed:
Actually modifying CommonParameters tests this time....

iSazonov commented:
@cmkb3 I wonder why do you change nuget configs? Do you use `Start-PSBuild -UseNuGetOrg`?

cmkb3 committed:
Update CommonParameters.Tests.ps1

cmkb3 committed:
Merge branch 'ProgressPreference_Fix' of https://github.com/cmkb3/PowerShell into ProgressPreference_Fix

cmkb3 commented:
> @cmkb3 I wonder why do you change nuget configs? Do you use `Start-PSBuild -UseNuGetOrg`?

Yes, I'm a git idiot and accidentally pushed those configs when trying to not do that. Tried to undo it and I've only made it worse so far, I'm afraid 😩

iSazonov commented:
@cmkb3 You can close the PR and open new one with clean commit.
